src/jarabe/desktop/favoriteslayout.py

```diff
@@ -172,7 +172,36 @@ def allocate_children(self, allocation, children):
         for child in children:
             if not self._grid.is_in_grid(child):
                 width, height = self._get_child_grid_size(child)
-                self._grid.add(child, width, height, None, None, locked=False)
+                x = y = None
+                if hasattr(child, "get_buddy"):
+                    md5hash = hashlib.md5(child.get_buddy().get_key())
+                    digest = abs(hash(md5hash.digest()))
+                    w = (self._width - (width * 3)) / 4
+                    h = (self._height - (height * 3)) / 4
+                    x = ((digest & 0xFFFFFFFF) % w) * ((digest >> 126) + 1)
+                    y = ((digest >> 32) % h) * (((digest >> 124) & 0b11) + 1)
+                elif hasattr(child, "get_first_ap"):
+                    digest = abs(child.get_first_ap().network_hash())
+                    w = (self._width - (width * 3)) / 4
+                    h = (self._height - (height * 3)) / 4
+                    x = ((digest & 0xFFFFFFFF) % w) * ((digest >> 126) + 1)
+                    y = ((digest >> 32) % h) * (((digest >> 124) & 0b11) + 1)
+                elif hasattr(child, "get_channel"):
+                    data = type(child).__name__ + str(child.get_channel())
+                    md5hash = hashlib.md5(data)
+                    digest = abs(hash(md5hash.digest()))
+                    w = (self._width - (width * 3)) / 4
+                    h = (self._height - (height * 3)) / 4
+                    x = ((digest & 0xFFFFFFFF) % w) * ((digest >> 126) + 1)
+                    y = ((digest >> 32) % h) * (((digest >> 124) & 0b11) + 1)
+                elif hasattr(child, "activity_id"):
+                    md5hash = hashlib.md5(str(child.activity_id))
+                    digest = abs(hash(md5hash.digest()))
+                    w = (self._width - (width * 3)) / 4
+                    h = (self._height - (height * 3)) / 4
+                    x = ((digest & 0xFFFFFFFF) % w) * ((digest >> 126) + 1)
+                    y = ((digest >> 32) % h) * (((digest >> 124) & 0b11) + 1)
+                self._grid.add(child, width, height, x, y, locked=False)
```

Comment:
there is a lot of duplicated code there!  Maybe instead of checking for each type of object, you could add a `get_positioning_data` function or something to each object.  That would return the data, whether it is the buddy key,network_has, channel or activity id.